File: homeostasis/solutions/sol.java
import java.io.*;
import java.util.*;

public class Solution {

    public static boolean areBracketsBalanced(String input) {
        Stack<Character> stack = new Stack<>();

        for (char bracket : input.toCharArray()) {
            if (bracket == '(' || bracket == '{' || bracket == '[') {
                stack.push(bracket);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.pop();
                if ((bracket == ')' && top != '(') || (bracket == '}' && top != '{') || (bracket == ']' && top != '[')) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }

    public static void main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        scanner.nextLine(); // Consume the newline character

        for (int i = 0; i < t; i++) {
            String brackets = scanner.nextLine();
            boolean answer = areBracketsBalanced(brackets);
            if (answer)
                System.out.println("True");
            else
                System.out.println("False");
        }

        scanner.close();
    }
}

File: homeostasis/solutions/sol.py
open_brackets = ['(', '{', '[']
close_brackets = [')', '}', ']']
bracket_conversion = {
    '(': ')',
    '{': '}',
    '[': ']'
}

def sol(brackets: str) -> bool:
    # Odd lengths cannot be valid
    if len(brackets) % 2 != 0:
        return False

    stack = []
    for bracket in brackets:
        if bracket in open_brackets:
            stack.append(bracket)
        else:
            if len(stack) == 0:
                return False
            if bracket_conversion[stack.pop()] != bracket:
                return False

    assert not stack
    return True


if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        answer = sol(input())
        print(answer)
